`runServer` fails with includes jar file size being above a certain threshold

The `runServer` task fails when the summed up jar file size of all includes is above a certain threshold.
I believe that threshold sits around 100MB.

The term "includes" here does not only mean direct includes, but also includes which come in through another mod via `modImplementation`.

To reproduce the issue, just add a lot of mods using `modImplementation` which in turn have a lot of large includes, and then use `runServer` to start a server in the dev environment.

This issue is only about the dev server, this issue does not affect production at all.

The exception is a very general one, seems like the service loader which searches for the GameProvider does not find anything:
```
[20:10:34] [ERROR] [FabricLoader/GameProvider]: Minecraft game provider couldn't locate the game! The game may be absent from the class path, lacks some expected files, suffers from jar corruption or is of an unsupported variety/version.
[20:10:34] [ERROR] [FabricLoader/]: Uncaught exception in thread "main"
java.lang.RuntimeException: Minecraft game provider couldn't locate the game! The game may be absent from the class path, lacks some expected files, suffers from jar corruption or is of an unsupported variety/version.
	at net.fabricmc.loader.impl.launch.knot.Knot.createGameProvider(Knot.java:214)
	at net.fabricmc.loader.impl.launch.knot.Knot.init(Knot.java:130)
	at net.fabricmc.loader.impl.launch.knot.Knot.launch(Knot.java:68)
	at net.fabricmc.loader.impl.launch.knot.KnotServer.main(KnotServer.java:23)
	at net.fabricmc.devlaunchinjector.Main.main(Main.java:86)
```

--- a/src/main/java/net/fabricmc/loom/task/AbstractRunTask.java
+++ b/src/main/java/net/fabricmc/loom/task/AbstractRunTask.java
@@ -25,20 +25,30 @@
 package net.fabricmc.loom.task;
 
 import java.io.File;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
 import java.util.ArrayList;
-import java.util.Collections;
 import java.util.List;
 import java.util.function.Function;
+import java.util.stream.Collectors;
 
 import org.gradle.api.Project;
+import org.gradle.api.file.ConfigurableFileCollection;
+import org.gradle.api.file.FileCollection;
 import org.gradle.api.specs.Spec;
 import org.gradle.api.tasks.JavaExec;
+import org.jetbrains.annotations.NotNull;
 
 import net.fabricmc.loom.configuration.ide.RunConfig;
 import net.fabricmc.loom.util.Constants;
 
 public abstract class AbstractRunTask extends JavaExec {
 	private final RunConfig config;
+	// We control the classpath, as we use a ArgFile to pass it over the command line: https://docs.oracle.com/javase/7/docs/technotes/tools/windows/javac.html#commandlineargfile
+	private final ConfigurableFileCollection classpath = getProject().getObjects().fileCollection();
 
 	public AbstractRunTask(Function<Project, RunConfig> configProvider) {
 		super();
@@ -46,6 +56,9 @@ public abstract class AbstractRunTask extends JavaExec {
 		this.config = configProvider.apply(getProject());
 
 		setClasspath(config.sourceSet.getRuntimeClasspath().filter(File::exists).filter(new LibraryFilter()));
+		// Pass an empty classpath to the super JavaExec.
+		super.setClasspath(getProject().files());
+
 		args(config.programArgs);
 		getMainClass().set(config.mainClass);
 	}
@@ -69,10 +82,44 @@ public abstract class AbstractRunTask extends JavaExec {
 
 	@Override
 	public List<String> getJvmArgs() {
-		List<String> superArgs = super.getJvmArgs();
-		List<String> args = new ArrayList<>(superArgs != null ? superArgs : Collections.emptyList());
+		final List<String> superArgs = super.getJvmArgs();
+		final List<String> args = new ArrayList<>();
+
+		final String content = "-classpath\n" + this.classpath.getFiles().stream()
+				.map(File::getAbsolutePath)
+				.collect(Collectors.joining(System.getProperty("path.separator")));
+
+		try {
+			final Path argsFile = Files.createTempFile("loom-classpath", ".args");
+			Files.writeString(argsFile, content, StandardCharsets.UTF_8);
+			args.add("@" + argsFile.toAbsolutePath());
+		} catch (IOException e) {
+			throw new UncheckedIOException("Failed to create classpath file", e);
+		}
+
+		if (superArgs != null) {
+			args.addAll(superArgs);
+		}
+
 		args.addAll(config.vmArgs);
 		return args;
+	}
+
+	@Override
+	public @NotNull JavaExec setClasspath(@NotNull FileCollection classpath) {
+		this.classpath.setFrom(classpath);
+		return this;
+	}
+
+	@Override
+	public @NotNull JavaExec classpath(Object @NotNull... paths) {
+		this.classpath.from(paths);
+		return this;
+	}
+
+	@Override
+	public @NotNull FileCollection getClasspath() {
+		return this.classpath;
 	}
 
 	private class LibraryFilter implements Spec<File> {

--- a/src/main/resources/eclipse_run_config_template.xml
+++ b/src/main/resources/eclipse_run_config_template.xml
@@ -16,4 +16,5 @@
     <stringAttribute key="org.eclipse.jdt.launching.PROJECT_ATTR" value="%ECLIPSE_PROJECT%"/>
     <stringAttribute key="org.eclipse.jdt.launching.VM_ARGUMENTS" value="%VM_ARGS%"/>
     <stringAttribute key="org.eclipse.jdt.launching.WORKING_DIRECTORY" value="${workspace_loc:%ECLIPSE_PROJECT%}/%RUN_DIRECTORY%"/>
+    <booleanAttribute key="org.eclipse.jdt.launching.ATTR_ATTR_USE_ARGFILE" value="true"/>
 </launchConfiguration>

--- a/src/main/resources/idea_run_config_template.xml
+++ b/src/main/resources/idea_run_config_template.xml
@@ -11,5 +11,6 @@
     <envs>
       %IDEA_ENV_VARS%
     </envs>
+    <shortenClasspath name="ARGS_FILE" />
   </configuration>
 </component>

--- a/src/test/groovy/net/fabricmc/loom/test/unit/IdeaClasspathModificationsTest.groovy
+++ b/src/test/groovy/net/fabricmc/loom/test/unit/IdeaClasspathModificationsTest.groovy
@@ -84,6 +84,7 @@ class IdeaClasspathModificationsTest extends Specification {
     <envs>
       %IDEA_ENV_VARS%
     </envs>
+    <shortenClasspath name="ARGS_FILE" />
   <classpathModifications><entry exclude="true" path="/path/to/file.jar"/></classpathModifications></configuration>
 </component>
 '''.trim()
@@ -103,6 +104,7 @@ class IdeaClasspathModificationsTest extends Specification {
     <envs>
       %IDEA_ENV_VARS%
     </envs>
+    <shortenClasspath name="ARGS_FILE" />
   <classpathModifications><entry exclude="true" path="/path/to/file.jar"/><entry exclude="true" path="/path/to/another.jar"/></classpathModifications></configuration>
 </component>
 '''.trim()

